### DISCUSSION TITLE
Memory and DOM element leaks in magic

### DISCUSSION THREAD

**Discussion by bep** [alpinejs/alpine#3283] [7. Bugs (archived)]
### Discussed in https://github.com/alpinejs/alpine/discussions/2139

<div type='discussions-op-text'>

<sup>Originally posted by **bep** September 28, 2021</sup>
This is related to #2126 -- but different enough to warrant its own issue.

This issue covers two different leaks, but they're closely related.

Both are demonstrated here with reference to possible patches:

https://alpine-testlab.netlify.app/leak2-magic/

## injectMagics holds a reference to el that never gets released

I have tested with this patch https://github.com/bep/alpine/commit/6c14dfc245d063382d80dfd5518bb8ad07edf017 -- which works for my simple test case, but I notice that `injectMagics` is also used elsewhere.

## magics' effects, e.g. in $watch, is never released

Which becomes a problem when they hold a reference to a DOM element (which they often do).

I have tested this patch: https://github.com/bep/alpine/commit/9dfe0c4b3bd5aeebd802e55910df79ed5ad07ea6

That works. However, it only works for `$watch` and it may make the making of custom `magics´ harder than it should.

</div>

**Comment by bep** [comment]
I can add that I have some additional and related issues so I would maybe hold off a little if someone wants to try to fix the above.

**Comment by calebporzio** [comment]
@bep - Ok, I think I fixed the "magic" function memory leaks caused by `$watch`

If you get a chance to check if there are any more leaks on the master branch that would be great!

Here's my code: https://github.com/alpinejs/alpine/commit/260679bec16f44c891b2c2082d020b53cc834eee

**Comment by bep** [comment]
@calebporzio thanks for that, I took the latest Alpine for a spin, and now the only patch I have to apply to get my "test cases" green is the one below:

https://github.com/bep/alpine/commit/91231f1ac33f3740dd06ba3ac23f29abe0eebb56

Note that I'm not totally sure if that it the best patch for the above, but it works for me.

**Comment by joshhanley** [comment]
This issue has been converted to a discussion as we are now using discussions to track bugs and ideas.
See this announcement for more details https://github.com/alpinejs/alpine/discussions/1423
